### PR TITLE
employee standing debtor and creditor transactions

### LINKED ISSUE #3510: Employee Standing Report should optionally include debtor transactions
The Employee Standing Report should optionally include the debtor transactions, preferably via a yes/no radio.

The use-case is this: a hospital administrator wants to know how much they are spending on an employee care.  Instead of locating first the patient associated with the employee, and then the patient standing report, they could simply combine both sets of information in the employee standing report.

### PR DESCRIPTION
**Pull request by jeremielodi**
closes #3510

### CODE REVIEW

**Comment by jeremielodi** [comment]
employye

File: server/controllers/finance/reports/financial.employee.handlebars [hunk 1]
```diff
@@ -6,20 +6,21 @@
...
-  <h5 class="text-uppercase"><strong>{{translate "FORM.LABELS.REGISTRATION_NUMBER"}} : {{ employee.code }}</strong></h5>
+  <h4 class="text-uppercase"><b>{{translate "FORM.LABELS.EMPLOYEE_NAME"}} :  {{ employee.display_name }}</b></h4>
+  <h4 class="text-uppercase"><b>{{translate "FORM.LABELS.REFERENCE"}} : {{ employee.reference }}</b></h4>
+  <h5 class="text-uppercase"><b>{{translate "FORM.LABELS.REGISTRATION_NUMBER"}} : {{ employee.code }}</b></h5>
   <h4 class="text-uppercase" style="background-color: #efefef;">
-    <strong>
-      {{translate aggregates.balanceText}} == {{debcred aggregates.balance metadata.enterprise.currency_id}}
-    </strong>
+    <b>
+      {{translate "FORM.LABELS.BALANCE"}}== {{debcred ./employeeBalance metadata.enterprise.currency_id}}
```
Comment (jniles) [inline server/controllers/finance/reports/financial.employee.handlebars:17]: :+1:

File: server/controllers/finance/reports/financial.employee.handlebars [hunk 2]
```diff
@@ -68,5 +69,58 @@
...
+          <th style="width: 15%;" class="text-center">{{translate "TABLE.COLUMNS.BALANCE" }}</th>
+        </tr>
+      </thead>
+      <tbody>
+        {{#each debtorTransactions}}
+          <tr>
+            <td>{{date this.trans_date}}</td>
+            <td class="text-right" title="{{this.document}}">{{this.document}}</td>
+            <td class="text-right" title="{{this.trans_id}}">{{this.trans_id}}</td>
+            <td class="text-uppercase" style="max-width : 200px; white-space : nowrap; overflow : hidden; text-overflow : ellipsis;" title="{{this.description}}">{{this.description}}</td>
```
Comment (jniles) [inline server/controllers/finance/reports/financial.employee.handlebars:94]: Nice work on the description as a title 👍

File: server/controllers/finance/reports/financial.employee.js [hunk 3]
```diff
@@ -42,22 +45,34 @@ function build(req, res, next) {
...
-    .catch(next)
-    .done();
+
+  // creditor information
+  // should addd something here
+  data.employee = await Employee.lookupEmployee(options.employee_uuid);
+
+  const { transactions, aggregates } = await Creditors.getFinancialActivity(data.employee.creditor_uuid);
+
+  aggregates.balanceText = aggregates.balance >= 0 ? 'FORM.LABELS.CREDIT_BALANCE' : 'FORM.LABELS.DEBIT_BALANCE';
```
Comment (jniles) [inline server/controllers/finance/reports/financial.employee.js]: I don't think this is used anymore.  Could you remove it?

File: server/controllers/finance/reports/financial.employee.js [hunk 4]
```diff
@@ -42,22 +45,34 @@ function build(req, res, next) {
   }
 
   const data = {};
+  const sql = `
+    SELECT BUID(p.debtor_uuid) as debtor_uuid
+    FROM patient p, employee em
```
Comment (jniles) [inline server/controllers/finance/reports/financial.employee.js]: Everywhere else we use `FROM patient p JOIN employee em ON p.uuid = em.xxx`.  In order to make it easier for new developers, could you use this syntax here?

File: server/controllers/finance/reports/financial.employee.js [hunk 5]
```diff
@@ -12,6 +12,9 @@ const ReportManager = require('../../../lib/ReportManager');
 
 const Employee = require('../../payroll/employees');
 const Creditors = require('../../finance/creditors');
+const Debitors = require('../../finance/debtors');
```
Comment (jniles) [inline server/controllers/finance/reports/financial.employee.js]: This should be called `Debtors` not `Debitors`.

File: server/controllers/finance/reports/financial.employee.js [hunk 6]
```diff
@@ -42,22 +45,34 @@ function build(req, res, next) {
...
+  // should addd something here
+  data.employee = await Employee.lookupEmployee(options.employee_uuid);
+
+  const { transactions, aggregates } = await Creditors.getFinancialActivity(data.employee.creditor_uuid);
+
+  aggregates.balanceText = aggregates.balance >= 0 ? 'FORM.LABELS.CREDIT_BALANCE' : 'FORM.LABELS.DEBIT_BALANCE';
+  _.extend(data, { transactions, aggregates });
+
+  // debtor informations
+  const patient = await db.one(sql, db.bid(options.employee_uuid));
```
Comment (jniles) [inline server/controllers/finance/reports/financial.employee.js]: The use of async/await is really good!

We can make this more efficient by regrouping the queries and running a few of them in parallel with `Promise.all()`. Like this:

```js
const { employee, patient } = await Promise.all([
  Employee.lookupEmployee(options.employee_uuid),
  db.one(sql, db.bid(options.employee_uuid)),
]);

// now we have both the patient and employee
data.employee = employee;

// get debtor/creditor information
const [creditorOperations, debtorOperations] = await Promise.all([
  Creditors.getFinancialActivity(data.employee.creditor_uuid),
  Debitors.getFinancialActivity(patient.debtor_uuid, true),
]);

const { transactions, aggregates } = creditorOperations;
// etc...
```

**Review by jniles — COMMENTED** [review]
Looks pretty good!  I've made a comment to increase performance, but otherwise, LGTM.

File: server/controllers/finance/reports/financial.employee.js [hunk 7]
```diff
@@ -41,23 +44,43 @@ function build(req, res, next) {
...
+    // get debtor/creditor information
+    const [creditorOperations, debtorOperations] = await Promise.all([
+      Creditors.getFinancialActivity(employee.creditor_uuid),
+      Debtors.getFinancialActivity(patient.debtor_uuid, true),
+    ]);
+
+    _.extend(data, {
+      employee,
+      creditorTransactions : creditorOperations.transactions,
+      creditorAggregates : creditorOperations.aggregates,
```
Comment (jniles) [inline server/controllers/finance/reports/financial.employee.js:70]: Nice naming 👍

**Review by jniles — APPROVED** [review]
LGTM.

bors r+

**Pull request merged**